File: java/leetcode/TrieImp.java
//Leetcode problem  208 Implement Trie(Prefix Tree)
//Solution written by Xuqiang Fang on 15 May, 2018

import java.util.Queue;
import java.util.LinkedList;

class Trie{
    /** Initialize your data structure here */
    public Trie(){
        root = new TrieNode(); 
    }
    
    /** Insert a word into the trie*/
    public void insert(String word){
        TrieNode p = root;
        for(int i=0; i<word.length(); ++i){
            int index = (int)(word.charAt(i) - 'a');
            if(p.children[index] == null){
                p.children[index] = new TrieNode();
            }
            p = p.children[index];
        }
        p.isWord = true;
    }

    /** Returns true if the word is in the trie*/
    public boolean search(String word){
        TrieNode node = find(word);
        return node != null && node.isWord;
    }

    /**Returns if there is any word in the trie that starts with the given prefix*/
    public boolean startsWith(String prefix){
        TrieNode node = find(prefix);
        return node != null;
    }

    private TrieNode find(String prefix){
        TrieNode p = root;
        for(int i=0; i<prefix.length(); ++i){
            int index = (int)(prefix.charAt(i) - 'a');
            if(p.children[index] == null)
                return null;
            p = p.children[index];
        }
        return p;
    }

    public void printTrie(){
        Queue<TrieNode> queue = new LinkedList<>();
        queue.offer(root);
        while(!queue.isEmpty()){
            TrieNode node = queue.poll();
            for(int i=0; i<26; ++i){
                if(node.children[i] != null){
                    TrieNode child = node.children[i];
                    queue.offer(child);
                    System.out.println((char)(i+'a'));
                }
            }
        }
    }

    class TrieNode{
        public TrieNode(){
            children = new TrieNode[26];
            isWord = false;
        } 
        public TrieNode[] children;
        public boolean isWord;
    }
    TrieNode root;
}

public class TrieImp{
	public static void main(String[] args){
        Trie t = new Trie();
        t.insert("hello");
        t.insert("world");
        t.insert("i");
        t.insert("love");
        t.insert("leetcode");
        System.out.println(t.search("leetcode"));
        System.out.println(t.startsWith("leet"));
        System.out.println(t.startsWith("leetfdj"));
        t.printTrie();
	}
}

File: cpp/leetcode/TrieImp.cpp
//Leetcode Problem No 208 Implement Trie (Prefix Tree)
//Solution written by Xuqiang Fang on 29 May, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class TrieNode{
public:
    TrieNode(bool b = false){
        is_word = b;
        memset(next, 0, sizeof(next));
    }
    TrieNode *next[26];
    bool is_word;
};
class Trie{
public:
    /** Initialize your data structure here. */
    Trie() {
        root = new TrieNode();
    }
    ~Trie(){
        clear(root);
    } 
    /** Inserts a word into the trie. */
    void insert(string word) {
        auto t = root;
        for(int i=0; i<word.length(); ++i){
            if(t->next[word[i]-'a'] == nullptr){
                t->next[word[i]-'a'] = new TrieNode();
            }
            t = t->next[word[i]-'a'];
        }
        t->is_word = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        auto t = find_string(word);
        if(t != nullptr && (t->is_word)){
            return true;
        }
        return false;    
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        auto t = find_string(prefix);
        if(t != nullptr){
            return true;
        }
        return false; 
    }
private:
    TrieNode *root;
    TrieNode *find_string(string word){
        auto temp = root;
        for(int i=0; i<word.length(); ++i){
            if(temp->next[word[i]-'a'] != nullptr){
                temp = temp->next[word[i]-'a'];
            }
            else{
                temp = nullptr;
                break;
            }
        }
        return temp;
    }
    void clear(TrieNode *root){
        for(int i=0; i<26; ++i){
            if(root -> next[i] != nullptr){
                clear(root->next[i]);
            }
        }
        delete root;
    }
};

int main(){
    Trie t;
    t.insert("abcd");
    cout << t.startsWith("abc") << endl;
    cout << t.search("abc") << endl;
    cout << t.search("abcd") << endl;
    return 0;
}
